operator/pkg/model/translation/ingress/shared_ingress.go

```diff
@@ -0,0 +1,225 @@
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/cilium/cilium/operator/pkg/model"
+	"github.com/cilium/cilium/operator/pkg/model/translation"
+	ciliumv2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+)
+
+const (
+	secureHost   = "secure"
+	insecureHost = "insecure"
+)
+
+var _ translation.Translator = (*SharedIngressTranslator)(nil)
+
+// SharedIngressTranslator is the translator from model to CiliumEnvoyConfig
+// for Ingress.
+//
+// This translator is used for shared LB mode.
+//   - only one instance of CiliumEnvoyConfig with two listeners (secure and
+//     in-secure).
+//   - no LB service and endpoint
+type SharedIngressTranslator struct {
+	name             string
+	namespace        string
+	secretsNamespace string
+	enforceHTTPs     bool
+}
+
+// NewSharedIngressTranslator returns a new translator for shared ingress mode.
+func NewSharedIngressTranslator(name, namespace, secretsNamespace string, enforceHTTPs bool) *SharedIngressTranslator {
+	return &SharedIngressTranslator{
+		name:             name,
+		namespace:        namespace,
+		secretsNamespace: secretsNamespace,
+		enforceHTTPs:     enforceHTTPs,
+	}
+}
+
+// Translate translates the model to CiliumEnvoyConfig.
+func (i *SharedIngressTranslator) Translate(model *model.Model) (*ciliumv2.CiliumEnvoyConfig, *corev1.Service, *corev1.Endpoints, error) {
+	cec := &ciliumv2.CiliumEnvoyConfig{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      i.name,
+			Namespace: i.namespace,
+		},
+	}
+
+	cec.Spec.BackendServices = i.getBackendServices(model)
+	cec.Spec.Services = i.getServices(model)
+	cec.Spec.Resources = i.getResources(model)
+
+	return cec, nil, nil, nil
+}
+
+func (i *SharedIngressTranslator) getBackendServices(m *model.Model) []*ciliumv2.Service {
+	var res []*ciliumv2.Service
+
+	for ns, v := range getNamespaceNamePortsMap(m) {
+		for name, ports := range v {
+			res = append(res, &ciliumv2.Service{
+				Name:      name,
+				Namespace: ns,
+				Ports:     ports,
+			})
+		}
+	}
+
+	// Make sure the result is sorted by namespace and name to avoid any
+	// nondeterministic behavior.
+	sort.Slice(res, func(i, j int) bool {
+		if res[i].Namespace != res[j].Namespace {
+			return res[i].Namespace < res[j].Namespace
+		}
+		if res[i].Name != res[j].Name {
+			return res[i].Name < res[j].Name
+		}
+		return res[i].Ports[0] < res[j].Ports[0]
+	})
+	return res
+}
+
+func (i *SharedIngressTranslator) getServices(_ *model.Model) []*ciliumv2.ServiceListener {
+	return []*ciliumv2.ServiceListener{
+		{
+			Listener:  fmt.Sprintf("%s-%s-listener", i.namespace, i.name),
+			Name:      i.name,
+			Namespace: i.namespace,
+		},
+	}
+}
+
+func (i *SharedIngressTranslator) getResources(m *model.Model) []ciliumv2.XDSResource {
+	var res []ciliumv2.XDSResource
+	res = append(res, i.getListener(m)...)
+	res = append(res, i.getRouteConfiguration(m)...)
+	res = append(res, i.getClusters(m)...)
```

Comment:
We may improve the allocation efficiency doing something like this:
```
	listener, routeConfig, clusters := i.getListener(m), i.getRouteConfiguration(m), i.getClusters(m)
	res := make([]ciliumv2.XDSResource, 0, len(listener)+len(routeConfig)+len(clusters))
	res = append(res, listener...)
	res = append(res, routeConfig...)
	res = append(res, clusters...)
```
But I see that the code is less elegant.
I don't know the usual size of `routeConfig` and `clusters` slices, feel free to discard this comment if we expect them to be small.